packages/expo-modules-autolinking/src/dependencies/CachedDependenciesLinker.ts

```diff
@@ -0,0 +1,139 @@
+import { loadConfigAsync } from '../reactNativeConfig/config';
+
+export interface CachedDependenciesSearchOptions {
+  excludeNames: Set<string>;
+  searchPaths: string[];
+}
+
+export interface CachedDependenciesLinker {
+  getOptionsForPlatform(platform: SupportedPlatform): Promise<CachedDependenciesSearchOptions>;
+  loadReactNativeProjectConfig(): Promise<RNConfigReactNativeProjectConfig | null>;
+  scanDependenciesFromRNProjectConfig(): Promise<ResolutionResult>;
+  scanDependenciesRecursively(): Promise<ResolutionResult>;
+  scanDependenciesInSearchPath(searchPath: string): Promise<ResolutionResult>;
+}
+
+export function makeCachedDependenciesLinker(params: {
+  projectRoot: string;
+}): CachedDependenciesLinker {
+  const linkingOptionsFactory = createLinkingOptionsFactory<SearchOptions>({
+    projectRoot: params.projectRoot,
+    platform: 'apple', // Placeholder value
+  });
+
+  let projectRoot: Promise<string> | undefined;
+  const getProjectRoot = () =>
+    projectRoot || (projectRoot = linkingOptionsFactory.getProjectRoot());
+
+  const dependenciesResultBySearchPath = new Map<string, Promise<ResolutionResult>>();
+  let reactNativeProjectConfig: Promise<RNConfigReactNativeProjectConfig | null> | undefined;
+  let reactNativeProjectConfigDependencies: Promise<ResolutionResult> | undefined;
+  let recursiveDependencies: Promise<ResolutionResult> | undefined;
+
+  return {
+    async getOptionsForPlatform(platform) {
+      const options = await linkingOptionsFactory.getPlatformOptions(platform);
+      return makeCachedDependenciesSearchOptions(options);
+    },
+    async loadReactNativeProjectConfig() {
+      if (reactNativeProjectConfig === undefined) {
+        reactNativeProjectConfig = loadConfigAsync<RNConfigReactNativeProjectConfig>(
+          await getProjectRoot()
+        );
+      }
+      return reactNativeProjectConfig;
+    },
+    async scanDependenciesFromRNProjectConfig() {
+      const reactNativeProjectConfig = await this.loadReactNativeProjectConfig();
+      return (
+        reactNativeProjectConfigDependencies ||
+        (reactNativeProjectConfigDependencies = scanDependenciesFromRNProjectConfig(
+          await getProjectRoot(),
+          reactNativeProjectConfig
+        ))
+      );
+    },
+    async scanDependenciesRecursively() {
+      return (
+        recursiveDependencies ||
+        (recursiveDependencies = scanDependenciesRecursively(await getProjectRoot()))
+      );
+    },
+    async scanDependenciesInSearchPath(searchPath: string) {
+      let result = dependenciesResultBySearchPath.get(searchPath);
+      if (!result) {
+        dependenciesResultBySearchPath.set(
+          searchPath,
+          (result = scanDependenciesInSearchPath(searchPath))
+        );
+      }
+      return result;
+    },
+  };
+}
+
+export async function scanDependencyResolutionsForPlatform(
+  linker: CachedDependenciesLinker,
+  platform: SupportedPlatform
+): Promise<ResolutionResult> {
```

Comment:
This is basically the main API to get a more efficient entrypoint into autolinking, so this is the main addition